### PR TITLE
Quality and diagnostic plots. 

### PR DESCRIPTION
**Pull request by felipeji**
This pull request includes the production of diagnostic plots during the data reduction process. The plots are saved as `.png` files in the `data_products/plots/` directory, instead of being interactively displayed during the reduction. The actual path and name of this directory can be configured in the` reduce_data.py` file in the `plot_dir` variable.

### CODE REVIEW

File: src/pywifes/wifes_calib.py [hunk 1]
```diff
@@ -960,54 +961,77 @@ def derive_wifes_telluric(
...
+            label=r"H$_2$O lines",
+        )
+
+        # for i, cube_fn in enumerate(cube_fn_list):
+        #     airmass = airmass_list[i]
+        #     wave, O2_ratio = O2_corrections[i]
+        #     H2O_ratio = H2O_corrections[i][1]
+
+        #     ax_bottom.plot(wave, O2_ratio**(1.0/(airmass**O2_power)), label=f'{cube_fn} O2')
+        #     ax_bottom.plot(wave, H2O_ratio**(1.0/(airmass**H2O_power)), label=f'{cube_fn} H2O')
```
Comment (CMartinezLombilla) [inline src/pywifes/wifes_calib.py:1018]: Is it useful somehow all this commented out code? If not, remove it, please.

File: src/pywifes/wifes_wsol.py [hunk 2]
```diff
@@ -879,7 +894,6 @@ def xcorr_shift_grid(slitlet_data, wave_guess, ref_interp):
     # pylab.plot(y_samp, all_stretches, 'bo')
     # pylab.plot(fval_y, stretch_fvals, 'g')
     # pylab.show()
```
Comment (CMartinezLombilla) [inline src/pywifes/wifes_wsol.py:896]: Is it useful the commented-out code above? If not, please remove it.

File: src/pywifes/wifes_wsol.py [hunk 3]
```diff
@@ -890,14 +904,10 @@ def xcorr_shift_grid(slitlet_data, wave_guess, ref_interp):
     # pylab.plot(y_samp[good_inds], all_shifts[good_inds], 'bo')
     # pylab.plot(fval_y, shift_fvals, 'g')
     # pylab.show()
```
Comment (CMartinezLombilla) [inline src/pywifes/wifes_wsol.py:906]: Same for this commented piece of code.

File: src/pywifes/wifes_wsol.py [hunk 4]
```diff
@@ -1568,62 +1560,56 @@ def _fit_optical_model(
...
+          if ((m.status == 5) and (fitcount < MAXFITS)):
+            print('mpfit needs more steps; repeating fit')
+          else:
+            fitdone = True
+
+            # if plot:
+            #   pl = numpy.asarray(m.params)[:om.nparams]
+            #   resid = om.errfunc(grating, pl, alphap, alls, ally, allx, allarcs)
+            #   om.plotResid(title,allx,ally,allarcs,resid)
+            #   om.plotFunc(title,allx,ally,allarcs,om.fitfunc(grating, pl,alphap,alls,ally,allx))
```
Comment (CMartinezLombilla) [inline src/pywifes/wifes_wsol.py:1612]: Same for this piece od commented-out code. Remove it if not useful.

File: src/pywifes/wifes_wsol.py [hunk 5]
```diff
@@ -1657,35 +1643,23 @@ def _fit_optical_model(
             resid = om.errfunc(
                 grating, pl[: om.nparams], pl[om.nparams :], alls, ally, allx, allarcs
             )
-            if doplot:
-                om.plotResid(title, allx, ally, allarcs, resid)
+        #   if (plot):
+        #     om.plotResid(title,allx,ally,allarcs,resid)
```
Comment (CMartinezLombilla) [inline src/pywifes/wifes_wsol.py:1647]: Same here.

**Review by CMartinezLombilla — APPROVED** [review]
Great job! Just some minor changes to clean up the codes.

File: src/pywifes/extract_spec.py [hunk 6]
```diff
@@ -21,11 +22,21 @@
 from pywifes.logger_config import custom_print
 import logging
 
+
 # Redirect print statements to logger
 logger = logging.getLogger("PyWiFeS")
+logger = logging.getLogger("PyWiFeS")
```
[the remark this answers is not in the document]
Comment (felipeji) [reply src/pywifes/extract_spec.py]: done!

File: src/pywifes/lacosmic.py [hunk 7]
```diff
@@ -15,9 +15,12 @@
 import logging
 
 
+# ------------------------------------------------------------------------
 # ------------------------------------------------------------------------
 # high-level functions to check if an observation is half-frame or N+S
 def _is_halfframe(hdus, data_hdu=0):
+    detsec = hdus[data_hdu].header["DETSEC"]
+    ystart = int(float(detsec.split(",")[1].split(":")[0]))
     detsec = hdus[data_hdu].header["DETSEC"]
```
[the remark this answers is not in the document]
Comment (felipeji) [reply src/pywifes/lacosmic.py:24]: done!

File: src/pywifes/lacosmic.py [hunk 8]
```diff
@@ -71,27 +74,26 @@ def lacos_spec_data(
...
-        blkdata = blkrep(subbed_data, 2, 2)
-        init_conv_data = scipy.signal.convolve2d(blkdata, laplace_kernel)[1:-1, 1:-1]
+        blkdata = blkrep(subbed_data,2,2)
+        init_conv_data = scipy.signal.convolve2d(blkdata, laplace_kernel)[1:-1,1:-1]
         init_conv_data[numpy.nonzero(init_conv_data <= 0.0)] = 0.0
         conv_data = blkavg(init_conv_data, 2, 2)
 
+        # ------------------------------------
+        conv_data = blkavg(init_conv_data, 2, 2)
+
```
[the remark this answers is not in the document]
Comment (felipeji) [reply src/pywifes/lacosmic.py:84]: done!

File: src/pywifes/lacosmic.py [hunk 9]
```diff
@@ -71,27 +74,26 @@ def lacos_spec_data(
...
         # ------------------------------------
         # step 4 - make noise model, create sigma map
         noise = (1.0 / gain) * ((gain * m5_model + rdnoise**2) ** 0.5)
+        noise = (1.0 / gain) * ((gain * m5_model + rdnoise**2) ** 0.5)
         noise_min = 0.00001
         noise[numpy.nonzero(noise <= noise_min)] = noise_min
         # div by 2 to correct convolution counting
         sigmap = conv_data / (2.0 * noise)
+        sigmap = conv_data / (2.0 * noise)
         # subtract large structure in sigma map
```
[the remark this answers is not in the document]
Comment (felipeji) [reply src/pywifes/lacosmic.py:94]: done!

File: src/pywifes/wifes_calib.py (hunk 1 above)
Comment (felipeji) [reply src/pywifes/wifes_calib.py:1018]: Yeah, this is coming from the refactorization of the plots logic. I'll leave it as it is for now.

File: src/pywifes/wifes_wsol.py (hunk 3 above)
Comment (felipeji) [reply src/pywifes/wifes_wsol.py:906]: I think I can reuse this piece of code for an extra plot for checking wavelength solution. I'll leave it for now.

File: src/pywifes/wifes_wsol.py (hunk 4 above)
Comment (felipeji) [reply src/pywifes/wifes_wsol.py:1612]: Same here. I'll back to it soon.

File: src/pywifes/lacosmic.py [hunk 10]
```diff
@@ -71,27 +74,26 @@ def lacos_spec_data(
...
         init_conv_data[numpy.nonzero(init_conv_data <= 0.0)] = 0.0
         conv_data = blkavg(init_conv_data, 2, 2)
 
+        # ------------------------------------
+        conv_data = blkavg(init_conv_data, 2, 2)
+
         # ------------------------------------
         # step 4 - make noise model, create sigma map
         noise = (1.0 / gain) * ((gain * m5_model + rdnoise**2) ** 0.5)
+        noise = (1.0 / gain) * ((gain * m5_model + rdnoise**2) ** 0.5)
```
[the remark this answers is not in the document]
Comment (felipeji) [reply src/pywifes/lacosmic.py:82]: done!

File: src/pywifes/optical_model.py [hunk 11]
```diff
@@ -639,7 +634,56 @@ def excludeAuto(lines, grating, bin_x, bin_y, resid, sigma, doplot, verbose):
...
       keepargs = np.logical_and(keepargs, np.logical_not(excludeargs))
 
-  if (doplot):
+  if (plot):
     plotResidKeep(allx,ally,allarcs,resid,keepargs)
 
   return lines[keepargs]
+
+
+import matplotlib.gridspec as gridspec
```
[the remark this answers is not in the document]
Comment (felipeji) [reply src/pywifes/optical_model.py:643]: Yeah, sorry! Done.

File: src/pywifes/pywifes.py [hunk 12]
```diff
@@ -1364,7 +1362,7 @@ def generate_wifes_bias_fit(
                 row_med[i] = numpy.mean(curr_col[good_inds])
             # row_med = numpy.median(curr_data, axis=0)
             bias_sub = row_med ** numpy.ones(numpy.shape(curr_data), dtype="d")
-            # Fred's update (bias fit) ------
+            # 's update (bias fit) ------
             # To remove the variations (some at least) along the
```
[the remark this answers is not in the document]
Comment (felipeji) [reply src/pywifes/pywifes.py:1366]: done

File: src/pywifes/pywifes.py [hunk 13]
```diff
@@ -1451,8 +1448,13 @@ def generate_wifes_bias_fit(
             )
         # Plot for test purposes ...
         if plot:
-            plt.figure()
+            plt.figure(1)
             plt.plot(linx, curr_data.mean(axis=0), "k-", label="raw bias", lw=2)
+
+            # Avoid ploting ouliers peaks
```
[the remark this answers is not in the document]
Comment (felipeji) [reply src/pywifes/pywifes.py:1454]: You should consider English is not my mother tongue. :) Done!

File: src/pywifes/pywifes.py [hunk 14]
```diff
@@ -2699,8 +2684,6 @@ def wifes_2dim_response(
                 alt_y = rect_spat_data[q, :]
                 norm_func = alt_y / (alt_flat_spec * next_normed_data[q, :])
                 alt_interp_spat[q, :] = norm_func
-                # print len(new_data)
-                # print len(norm_func)
                 # norm_func = (alt_y / (
                 #    alt_flat_spec))
                 # sp1.plot(norm_func[500:-200])
```
[the remark this answers is not in the document]
Comment (felipeji) [reply src/pywifes/pywifes.py:2689]: Prints without '()' . So something that remains from the previous version. Removed!

File: src/pywifes/pywifes.py [hunk 15]
```diff
@@ -10,7 +10,6 @@
 import os
 import scipy.ndimage as ndimage
 import scipy.interpolate as interp
-import pylab
 import matplotlib.pyplot as plt
 import pylab
```
[the remark this answers is not in the document]
Comment (felipeji) [reply src/pywifes/pywifes.py:14]: It is in use. I leave it.

File: src/pywifes/wifes_calib.py [hunk 16]
```diff
@@ -754,11 +760,12 @@ def derive_wifes_calibration(
...
-        pylab.show()
+
+        plot_name = "flux_calibration_solution.png"
+        plot_path = os.path.join(plot_dir, plot_name)
+        pylab.savefig(plot_path, dpi=300)
+        pylab.close()
+
     # Fred's update ... now, careful, because that's dirty ...
     # the function does not always return the same thing !
     # SAVE IN THE PICKLE FILE THE WAVELENGTH AND CALIB FVAL ARRAYS
```
[the remark this answers is not in the document]
Comment (felipeji) [reply src/pywifes/wifes_calib.py:771]: done!

File: src/pywifes/wifes_calib.py [hunk 17]
```diff
@@ -490,7 +493,6 @@ def derive_wifes_calibration(
         f = pyfits.open(cube_fn_list[i])
         cube_hdr = f[first].header
         f.close()
-
         # ------------------------------------
         # figure out which star it is
         # NEW VERSION 0.7.0: smart star name lookup!
```
[the remark this answers is not in the document]
Comment (felipeji) [reply src/pywifes/wifes_calib.py:498]: done

File: src/pywifes/wifes_calib.py [hunk 18]
```diff
@@ -796,6 +803,9 @@ def calibrate_wifes_cube(inimg, outimg, calib_fn, mode="pywifes", extinction_fn=
     wave0 = f3[first].header["CRVAL1"]
     dwave = f3[first].header["CDELT1"]
     exptime = f3[first].header["EXPTIME"]
+    wave0 = f3[first].header["CRVAL1"]
+    dwave = f3[first].header["CDELT1"]
+    exptime = f3[first].header["EXPTIME"]
```
[the remark this answers is not in the document]
Comment (felipeji) [reply src/pywifes/wifes_calib.py:808]: done!

File: src/pywifes/wifes_wsol.py (hunk 2 above)
Comment (felipeji) [reply src/pywifes/wifes_wsol.py:895]: Want to keep it. Probably will reused soon.

**Pull request merged**